src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Map.cs

```diff
@@ -60,5 +63,104 @@ public void ReadEndMap()
                 AdvanceDataItemCounters();
             }
         }
+
+        //
+        // Map decoding conformance
+        //
+
+        private void HandleMapKeyAdded()
+        {
+            Debug.Assert(_currentKeyOffset != null && _curentItemIsKey);
+
+            (int offset, int length) currentKeyRange = (_currentKeyOffset.Value, _bytesRead - _currentKeyOffset.Value);
+
+            if (CborConformanceLevelHelpers.RequiresSortedKeys(ConformanceLevel))
+            {
+                ValidateSortedKeyEncoding(currentKeyRange);
+            }
+            else if (CborConformanceLevelHelpers.RequiresUniqueKeys(ConformanceLevel))
+            {
+                ValidateKeyUniqueness(currentKeyRange);
+            }
+
+            _curentItemIsKey = false;
+        }
+
+        private void HandleMapValueAdded()
+        {
+            Debug.Assert(_currentKeyOffset != null && !_curentItemIsKey);
+
+            _currentKeyOffset = _bytesRead;
+            _curentItemIsKey = true;
+        }
+
+        private void ValidateSortedKeyEncoding((int offset, int length) currentKeyRange)
+        {
+            Debug.Assert(_currentKeyOffset != null);
+
+            if (_previousKeyRange != null)
+            {
+                (int offset, int length) previousKeyRange = _previousKeyRange.Value;
+
+                ReadOnlySpan<byte> originalBuffer = _originalBuffer.Span;
+                ReadOnlySpan<byte> previousKeyEncoding = originalBuffer.Slice(previousKeyRange.offset, previousKeyRange.length);
+                ReadOnlySpan<byte> currentKeyEncoding = originalBuffer.Slice(currentKeyRange.offset, currentKeyRange.length);
+
+                int cmp = CborConformanceLevelHelpers.CompareEncodings(previousKeyEncoding, currentKeyEncoding, ConformanceLevel);
+                if (cmp > 0)
+                {
+                    throw new FormatException("CBOR map keys are not in sorted encoding order.");
+                }
+                else if (cmp == 0 && CborConformanceLevelHelpers.RequiresUniqueKeys(ConformanceLevel))
+                {
+                    throw new FormatException("CBOR map contains duplicate keys.");
+                }
+            }
+
+            _previousKeyRange = currentKeyRange;
+        }
+
+        private void ValidateKeyUniqueness((int offset, int length) currentKeyRange)
+        {
+            Debug.Assert(_currentKeyOffset != null);
+
+            SortedSet<(int offset, int length)> previousKeys = GetPreviousKeyRanges();
+
+            if (!previousKeys.Add(currentKeyRange))
+            {
+                throw new FormatException("CBOR map contains duplicate keys.");
+            }
+        }
+
+        private SortedSet<(int offset, int length)> GetPreviousKeyRanges()
+        {
+            if (_previousKeyRanges == null)
+            {
+                _keyComparer ??= new KeyEncodingComparer(this);
+                _previousKeyRanges = new SortedSet<(int offset, int length)>(_keyComparer);
```

Comment:
On the reader side this is only used for key uniqueness in unsorted conformance levels, so a HashSet might have been preferable. However I couldn't find a good enough hashcode implementation for spans.

Reply:
You should use Marvin for that
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/Common/src/System/Marvin.cs

that's what we use (or should use) everywhere

Reply:
If you've got `span<char>` and this will only be 3.1 or later than you get use the `string.GetHashCode(ReadOnlySpan<char>,StringComparison>` which was added in 3.1 and ends up with Marvin. If you've got anything other than chars I'm not sure, is this a usecase for hashcode combine?

Reply:
TIL, thanks @krwq!

@Wraith2 In this case it's comparing spans of bytes.

Reply:
@GrabYourPitchforks do you perhaps know if we got any equivalents for byte? Would be better to avoid generating random seed again for this library